src/hx711.toit

```diff
@@ -0,0 +1,178 @@
+// Copyright (C) 2021 Toitware ApS.  All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the LICENSE file.
+
+import gpio
+import uart
+
+/**
+Driver for the HX711 Analog-to-Digital converter.
+*/
+
+/**
+An argument to the constructor of $Hx711.
+Valid instances are $Hx711.CHANNEL_A_GAIN_128, $Hx711.CHANNEL_B_GAIN_32, or
+  $Hx711.CHANNEL_A_GAIN_64.
+*/
+class Hx711Input:
+  pulses/int
+
+  constructor.private_ .pulses:
+
+/**
+Driver for the HX711 Analog-to-Digital converter.
+Uses a UART to form the correct shaped pulses.
+This driver does not currently support power-down mode for the chip.
+*/
+class Hx711:
+  data_/gpio.Pin := ?
+  uart_port_/uart.Port := ?
+  input_/Hx711Input? := null
+
+  /// Read from channel A with a gain of 128. 
+  static CHANNEL_A_GAIN_128 ::= Hx711Input.private_ 25
+  /// Read from channel B with a gain of 32. 
+  static CHANNEL_B_GAIN_32  ::= Hx711Input.private_ 26
+  /// Read from channel A with a gain of 64. 
+  static CHANNEL_A_GAIN_64  ::= Hx711Input.private_ 27
+
+  /**
+  The $clock pin is used as an output pin, connected to the UART hardware.
+  The $data pin is used as an input GPIO pin.
+  */
+  constructor --clock/int=17 --data/int=16:
+    data_ = gpio.Pin.in 16
+    clock_pin := gpio.Pin.out clock
+    uart_port_ = uart.Port
+      --tx=clock_pin
+      --rx=null
+      // This common baud rate gives an 8.6us pulse for each start bit, which is
+      // in the range of 0.1us to 50us that the HX711 requires.
+      --baud_rate=115200
+      // We want the idle state to be low and the start bit to be high.
+      --invert_tx=true
+
+  // The actual pattern on the UART TX pin will be a start bit (high,
+  // 8.6us), eight data bits (low, since the TX pin is inverted) and a stop bit
+  // (low).
+  SINGLE_PULSE_ := #[0xff]
+
+  /**
+  Returns a value between -1.0 and 1.0.
+  May also return -Infinity or Infinity if the measurement is out of range.
+  For $input pass $Hx711.CHANNEL_A_GAIN_128, $Hx711.CHANNEL_B_GAIN_32, or
+    $Hx711.CHANNEL_A_GAIN_64.
+  */
+  get input/Hx711Input -> float:
+    if input_ != input:
+      // If the input setting is wrong we do a dummy read to set the input
+      // correctly.
+      data_.wait_for 0
+      input.pulses.repeat:
+        uart_port_.write SINGLE_PULSE_
+        sleep --ms=1
+      input_ = input
+
+    // Wait for a sample to be ready.
+    data_.wait_for 0
+
+    measurement := 0
+
+    24.repeat: 
+      uart_port_.write SINGLE_PULSE_
+      sleep --ms=1
```

Comment:
Why to you sleep 1 ms here?
Is it because the UART could be buffered? Are we sure that the UART has written with the sleep?
Is there maybe a `uart_port_.flush` which would be better?

Reply:
I think the uart write does not return until it is all written, but the HX711 needs a non-zero time to put the next bit on the data line.